### PR TITLE
feat/gzil

### PR DESCRIPTION
**Pull request by renlulu**
** Don't merge, only for pre review purpose **

This PR introduces integration with gzils, (after discussion) there is only one chain call would happen now:

1. Claim rewards => Mint gzil

Other changes:

2. Modify gzli contract to adopt proxy
3. Make proxy support gzil contract
4. Add a new field `maxcommchangerate` to limit commission change
5. Add a new field `maxcommrate `to make sure the max commission would not exceed 100%
6. Add a new transition `UpdateContractAddr` to update proxy address and gzil address

### CODE REVIEW

File: contracts/ssnlist.scilla [hunk 1]
```diff
@@ -167,7 +183,8 @@ let make_error =
 (***************************************************)
 contract SSNList(
   init_admin : ByStr20,
-  proxy_address : ByStr20
+  proxy_address : ByStr20,
+  bzil_address : ByStr20
```
[the remark this answers is not in the document]
Comment (renlulu) [reply contracts/ssnlist.scilla:187]: Both proxy and bzil?

File: contracts/ssnlist.scilla [hunk 2]
```diff
@@ -118,6 +118,18 @@ let option_add =
...
+  fun (new : Uint128) =>
+    let a = builtin lt old new in
+    let b = match a with
+    | True =>
+      builtin sub new old
+    | False =>
+      builtin sub old new
+    end
+    in 
+    b
```
Comment (anton-trunov) [inline contracts/ssnlist.scilla:132]:
```suggestion
    match a with
    | True =>
      builtin sub new old
    | False =>
      builtin sub old new
    end
```

File: contracts/proxy.scilla [hunk 3]
```diff
@@ -56,29 +48,38 @@ transition ChangeProxyAdmin(newAdmin : ByStr20)
       event e
     end
 end
-
+transition UpgradeBzilContract(newAddress: ByStr20)
```
Comment (bb111189) [inline contracts/proxy.scilla:51]:
```suggestion
transition UpgradeGZILContract(newAddress: ByStr20)
```

File: contracts/proxy.scilla [hunk 4]
```diff
@@ -56,29 +48,38 @@ transition ChangeProxyAdmin(newAdmin : ByStr20)
...
+    currentAdmin <- admin;
+    isAdmin = builtin eq currentAdmin _sender;
+    match isAdmin with
+    | True =>
+      gzil_contract := newAddress;
+      e = {_eventname : "Gzil Upgraded"; gzil_address : newAddress};
+      event e
+    | False =>
+      e = {_eventname : "UpgradeTo FailedNotAdmin"; gzil_address : newAddress};
```
Comment (bb111189) [inline contracts/proxy.scilla:60]:
```suggestion
      e = {_eventname : "UpgradeGZILContract FailedNotAdmin"; gzil_address : newAddress};
```

File: contracts/proxy.scilla [hunk 5]
```diff
@@ -215,52 +214,91 @@ transition AddFunds()
...
     send msgs
 end
-
 transition PopulateDepositAmtForDeleg(deleg_addr : ByStr20, ssn_addr : ByStr20, amt : Uint128)
     current_impl <- implementation;
     msg = {_tag : "PopulateDepositAmtForDeleg"; _recipient : current_impl; _amount : zero; ssn_addr : ssn_addr; deleg_addr : deleg_addr; amt : amt; initiator: _sender};
     msgs = one_msg msg;
     send msgs
+end
+(* Transitions will be sent to bzil implemetion *)
```
Comment (bb111189) [inline contracts/proxy.scilla:259]:
```suggestion
(* Transitions will be sent to gZIL implemetion *)
```

File: contracts/ssnlist.scilla [hunk 6]
```diff
@@ -139,6 +151,9 @@ type Error =
   | DelegHasNoSufficientAmt
   | SSNNoComm
   | DelegStakeNotEnough
+  | BzilValidationFailed
```
Comment (bb111189) [inline contracts/ssnlist.scilla:154]:
```suggestion
  | GZILValidationFailed
```

File: contracts/ssnlist.scilla [hunk 7]
```diff
@@ -159,6 +174,9 @@ let make_error =
       | DelegHasNoSufficientAmt => Int32 -14
       | SSNNoComm => Int32 -15
       | DelegStakeNotEnough => Int32 -16
+      | BzilValidationFailed => Int32 -17
```
Comment (bb111189) [inline contracts/ssnlist.scilla:177]:
```suggestion
      | GZILValidationFailed => Int32 -17
```

File: contracts/ssnlist.scilla [hunk 8]
```diff
@@ -462,6 +508,13 @@ procedure CalcStakeDelegPerCycle (tmp_arg : TmpArg)
     end
   end
 end
+procedure MintCall(recipient: ByStr20, amount: Uint128)
+  addr <- gziladdr;
+  msg_to_bzil = {_tag : "Mint"; _recipient : addr; _amount : uint128_zero;
```
Comment (bb111189) [inline contracts/ssnlist.scilla:513]:
```suggestion
  msg_to_gzil = {_tag : "Mint"; _recipient : addr; _amount : uint128_zero; 
```

File: contracts/ssnlist.scilla [hunk 9]
```diff
@@ -462,6 +508,13 @@ procedure CalcStakeDelegPerCycle (tmp_arg : TmpArg)
     end
   end
 end
+procedure MintCall(recipient: ByStr20, amount: Uint128)
+  addr <- gziladdr;
+  msg_to_bzil = {_tag : "Mint"; _recipient : addr; _amount : uint128_zero; 
+                  recipient : recipient; amount : amount};
+  msgs = one_msg msg_to_bzil;
```
Comment (bb111189) [inline contracts/ssnlist.scilla:515]:
```suggestion
  msgs = one_msg msg_to_gzil;
```

**Review by bb111189 — CHANGES_REQUESTED** [review]

File: contracts/ssnlist.scilla [hunk 10]
```diff
@@ -139,6 +149,9 @@ type Error =
   | DelegHasNoSufficientAmt
   | SSNNoComm
   | DelegStakeNotEnough
+  | GZILValidationFailed
+  | ExceedMaxChangeRate
+  | ExceedMaxRate
```
Comment (bb111189) [inline contracts/ssnlist.scilla:154]:
```suggestion
  | ExceedMaxCommRate
```

File: contracts/ssnlist.scilla [hunk 11]
```diff
@@ -159,6 +172,9 @@ let make_error =
       | DelegHasNoSufficientAmt => Int32 -14
       | SSNNoComm => Int32 -15
       | DelegStakeNotEnough => Int32 -16
+      | GZILValidationFailed => Int32 -17
+      | ExceedMaxChangeRate => Int32 -18
+      | ExceedMaxRate => Int32 -19
```
Comment (bb111189) [inline contracts/ssnlist.scilla:177]:
```suggestion
      | ExceedMaxCommRate => Int32 -19
```

File: contracts/ssnlist.scilla [hunk 12]
```diff
@@ -221,6 +243,28 @@ procedure ThrowError(err : Error)
   e = make_error err;
   throw e
 end
+procedure ValidateRate(rate: Uint128)
+  max_rate <- maxcommrate;
+  validate = uint128_le rate max_rate;
+  match validate with
+  | True =>
+  | False =>
+    e = ExceedMaxRate;
```
Comment (bb111189) [inline contracts/ssnlist.scilla:252]:
```suggestion
    e = ExceedMaxCommRate;
```

**Review by bb111189 — COMMENTED** [review]
Make the error code abit more verbose

**Review by bb111189 — APPROVED** [review]

**Pull request merged**